src/headless/converse-muc.js

```diff
@@ -941,74 +942,83 @@ converse.plugins.add('converse-muc', {
+            getOccupantByNickname (nick) {
+                return this.occupants.findOccupant({ nick });
+            },
 
-                if (occupant) {
-                    uri = occupant.get('jid') || `${this.get('jid')}/${occupant.get('nick')}`;
-                } else if (nicknames.includes(longest_match)) {
-                    // TODO: show a warning to the user that the person is not currently in the chat
-                    uri = `${this.get('jid')}/${longest_match}`;
-                } else {
-                    return;
+            parseTextForReferences (original_message) {
+                if (!original_message) return ['', []];
+                const findRegexInMessage = p.matchRegexInText(original_message);
+                const mentions = findRegexInMessage(p.mention_regex);
+                if (!mentions) return [original_message, []];
+
+                const known_nicknames = this.getAllKnownNicknames();
+                const getMatchesForNickRegex = nick_regex => [...findRegexInMessage(nick_regex)];
+                const getMatchesIndexesForNick = nick => {
+                    const regex = RegExp(`(${nick}(?:[,.!?\\s]|$))`, 'ig');
+                    return getMatchesForNickRegex(regex).map(match => match.index);
                 }
-                const obj = {
-                    'begin': index,
-                    'end': index + longest_match.length,
-                    'value': longest_match,
-                    'type': 'mention',
-                    'uri': encodeURI(`xmpp:${uri}`)
+                const makeObjectFromNick = nick => ({ nick: `@${nick}` });
+                const addMatchesForNick = obj => ({
+                  ...obj,
+                  matches: getMatchesIndexesForNick(obj.nick),
+                });
+                const isStringPrecededBySpace = p.isStringPrecededBySpace(original_message);
+                const isValidMentionByIndex = mention_index =>
+                    mention_index == 0 || isStringPrecededBySpace(mention_index);
+                const toIndexedNicknames = (matches_by_index, match) => {
+                  const matches = match.matches.reduce((matches, i) => {
+                    const existing_match_in_index = matches_by_index[i];
+                    if ((
+                        (existing_match_in_index && match.nick.length > existing_match_in_index.length)
+                        || !existing_match_in_index) && isValidMentionByIndex(i)
+                    ) {
+                        matches[i] = match.nick;
+                    }
+                    return matches;
+                  }, {});
+                  return { ...matches_by_index, ...matches };
+                };
+                
+                const mentions_by_index = known_nicknames
+                    .map(makeObjectFromNick)
+                    .map(addMatchesForNick)
+                    .reduce(toIndexedNicknames, {}) 
```

Comment:
I've been thinking about this PR tonight, and this code in particular makes me nervous.

In some large MUCs, you can have thousands of users, which means we would be running thousands of regexes every single time someone sends a message, which I think might become noticeable to the user as small freeze-ups. 

Additionally, I've been thinking about the fact that we add mentions for nicknames that aren't prefixed with an `@` symbol. Slack for  example doesn't highlight a nickname if there's no `@` in front of it, and I think that makes sense, because sometimes you want to refer to someone without disturbing them, and sometimes someone has a very generic nickname and you're not actually referring to them at all.

I think we could kill two birds with one stone here... if we update the code to only add mentions if the nickname is preceded by an `@` symbol, and to then to stop matching as soon as we have as many matching nicknames as `@` symbols in the message. Similarly, we can avoid running the regexes entirely if there aren't any `@` symbols in the message. That way we avoid some unnecessary regex matching.

What do you think @afuggini?

Reply:
Let me reflect on this and get back to you.

Also FYI I still need to add a test.